### PR TITLE
[WIP] add init version of convert pb to IPU style tool

### PR DESCRIPTION
**Pull request by youbeny**
add init version of convert pb to IPU style tool, use the following command
```
   ./bin/analyzer ./model.pb -convert-to-ipu-graphdef
```
will generate a pb file named "converted_model.pb" in current directory

### CODE REVIEW

File: include/halo/utils/cl_options.h [hunk 1]
```diff
@@ -49,6 +49,14 @@ static llvm::cl::opt<bool> PrintAnalysisReport(
     "print-analysis-report", llvm::cl::desc("Print analysis report"),
     llvm::cl::init(false));
 
+static llvm::cl::opt<bool> ConvertToIpuGraphDef(
```
Comment (weimingzha0) [inline include/halo/utils/cl_options.h:52]: These options are only for the "analyzer" tool. 
Shall we move them to "analyzer/driver.cc" ?

File: lib/parser/tensorflow/tf_parser.cc [hunk 2]
```diff
@@ -76,6 +81,64 @@ void TFParser::Init(BasicBlock* bb, Function* function,
   opts_ = opts;
 }
 
+Status TFParser::SetAttributes(tensorflow::NodeDef* cur_node) {
```
Comment (weimingzha0) [inline lib/parser/tensorflow/tf_parser.cc:84]: Change to "SetIPUAttributes()" ?

File: lib/parser/tensorflow/tf_parser.cc [hunk 3]
```diff
@@ -76,6 +81,64 @@ void TFParser::Init(BasicBlock* bb, Function* function,
...
+
+  // Serialization ipu format graphdef
+  std::fstream ofs(opts_.output_graphdef_filename,
+                   std::ios::out | std::ios::binary);
+  HLCHECK(!ofs.fail());
+  if (!graph_def.SerializeToOstream(&ofs)) {
+    LOG(ERROR) << "Serialization output pb file error";
+    return Status::ASSERTION;
+  }
+#if 0
```
Comment (weimingzha0) [inline lib/parser/tensorflow/tf_parser.cc:128]: check if file name extension is ".pbtxt" ?

File: lib/parser/tensorflow/tf_parser.cc (hunk 2 above)
Comment (weimingzha0) [reply lib/parser/tensorflow/tf_parser.cc:84]: Can we move this code to a separate file? We can have a TFParser subclass like 
class IPUConverter : TFParser() {
...
}

**Review by weimingzha0 — COMMENTED** [review]
Can we move the

File: lib/parser/tensorflow/tf_parser.cc (hunk 2 above)
Comment (youbeny) [reply lib/parser/tensorflow/tf_parser.cc]: tf_parser class depends on tf_convert.h.inc, IPUConvert cannot be decoupled cleanly, need further refactor

File: include/halo/utils/cl_options.h (hunk 1 above)
Comment (youbeny) [reply include/halo/utils/cl_options.h]: done

File: lib/parser/tensorflow/tf_parser.cc [hunk 4]
```diff
@@ -870,6 +870,68 @@ Status TFParser::ConvertDummyNode(const tensorflow::NodeDef& node_def) {
...
+  }
+
+  // Serialization ipu format graphdef
+  std::fstream ofs(filename, std::ios::out | std::ios::binary);
+  HLCHECK(!ofs.fail());
+  if (!graph_def.SerializeToOstream(&ofs)) {
+    LOG(ERROR) << "Serialization output pb file error";
+    return Status::ASSERTION;
+  }
+#if 0
```
Comment (weimingzha0) [inline lib/parser/tensorflow/tf_parser.cc:921]: you can check filename

**Review by weimingzha0 — APPROVED** [review]

**Pull request merged**